### ISSUE TITLE
Training Disparity Decoder instead of Depth

### ISSUE THREAD

**Issue by shreyas23** [nianticlabs/monodepth2#277]
Hi!

Would it be an equivalent problem to instead estimate disparity? In regards to the stereo training setup, this would entail warping the pixel coordinates according to the estimated disparity and converting disparity to depth disparity = (focal * baseline)/disp for the temporal frames instead of estimating inverse depth and using the stereo extrinsics. 

I've adapted the codebase to train this way, however I'm quickly getting instability in the network and NaNs in the gradients for the disparity decoder where the average sigmoid outputs for each scale's output very quickly become 1.0 (within a few steps). For reference, I've initialized the network with kaiming normal and have switched to leaky relu activations. 

Is there motivation to use inverse depth over disparity?

Also, why use a simplified version of the stereo extrinsics instead of explicitly calculating it according to the setup in the original KITTI paper?

Best,
Shreyas Niradi

**Comment by mrharicot** [comment]
Hi,

Disparity makes sense when training in stereo mode only, as it is the (typically horizontal) displacement between pixels of both images in the stereo pair. There is no notion of disparity for general camera configurations such as monocular training of kitti.
monodepth2 actually predicts inverse depth which is essentially disparity up to an affine transform (scale + bias).

Hope this helps!

**Comment by shreyas23** [comment]
Hi,

Yep, thanks! I'll have to figure out why I'm getting those NaNs, but that's a different issue likely.

May I also ask why you normalize the intrinsics matrix? It looks like the Project3D and BackprojectDepth modules are working with un-normalized pixel coordinates? Thanks in advance!

Best,
Shreyas

**Comment by shreyas23** [comment]
Ah shoot, I see that it's scaled back up according to the image scale later on in the dataloader. Thanks for the quick answer earlier, it's much appreciated!
